### PR TITLE
Ben/explain

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

File: finetune/base_models/__init__.py [hunk 1]
```diff
@@ -24,8 +24,8 @@ def get_encoder(cls, **kwargs):
         return cls.encoder(**kwargs)
 
     @classmethod
-    def get_featurizer(cls, X, encoder, config, train=False, reuse=None):
```
Comment (benleetownsend) [inline finetune/base_models/__init__.py]: This needs to be updated for the other featurizers

File: finetune/base_models/gpt/encoder.py [hunk 2]
```diff
@@ -188,31 +188,36 @@ def decode(self, ids):
         return "".join([self.decoder.get(word_idx, '<unk>') for word_idx in ids]).replace("</w>", " ")
 
 
-def to_spacy_attn(attn, tokens, token_starts, token_ends):
+def to_spacy_attn(attn, tokens, token_starts, token_ends, normalize=True):
```
Comment (benleetownsend) [inline finetune/base_models/gpt/encoder.py]: Change name?

File: finetune/base_models/gpt/encoder.py [hunk 3]
```diff
@@ -188,31 +188,36 @@ def decode(self, ids):
...
+        spacy_attn = spacy_attn/sum(spacy_attn)
     return {
         'attention_weights': spacy_attn,
         'token_starts': spacy_token_starts,
         'token_ends': spacy_token_ends
     }
 
 
-def finetune_to_indico_attention_weights(raw_texts, attn_weights, encoder):
+def finetune_to_indico_attention_weights(raw_texts, attn_weights, encoder, attention=True):
```
Comment (benleetownsend) [inline finetune/base_models/gpt/encoder.py]: change name?

Comment (benleetownsend) [reply finetune/base_models/gpt/encoder.py]: Probably not as this is in the public API??

File: finetune/base_models/gpt/featurizer.py [hunk 4]
```diff
@@ -34,15 +34,34 @@ def mask_attn_weights(w):
     return w
 
 
-def attn_weights(q, k, v, attn_pdrop, train=False, scale=False, mask=True):
+def explain_mask_attn_weights(w, lengths):
```
Comment (benleetownsend) [inline finetune/base_models/gpt/featurizer.py]: lengths and length_mask are unused now. Remove.

File: finetune/base_models/gpt/featurizer.py [hunk 5]
```diff
@@ -92,12 +112,13 @@ def multihead_qkv(x, n_state, n_head, train):
     return q, k, v
 
 
-def attn(x, scope, n_state, n_head, resid_pdrop, attn_pdrop, train=False, scale=False, mask=True):
+def attn(x, scope, n_state, n_head, resid_pdrop, attn_pdrop, train=False, scale=False, mask=True, explain=False, clf_pos=None):
```
Comment (benleetownsend) [inline finetune/base_models/gpt/featurizer.py]: remove clf_pos, it is unused now.

File: finetune/base_models/gpt/featurizer.py [hunk 6]
```diff
@@ -118,10 +139,12 @@ def mlp(x, scope, n_state, act_fn, resid_pdrop, train=False):
         return h2
 
 
-def block(x, n_head, act_fn, resid_pdrop, attn_pdrop, scope, train=False, scale=False):
+def block(x, n_head, act_fn, resid_pdrop, attn_pdrop, scope, train=False, scale=False, explain=False, clf_pos=None):
```
Comment (benleetownsend) [inline finetune/base_models/gpt/featurizer.py]: remove clf_pos , this is unused

File: finetune/base_models/gpt/featurizer.py [hunk 7]
```diff
@@ -164,6 +187,16 @@ def gpt_featurizer(X, encoder, config, train=False, reuse=None):
 
         X = tf.reshape(X, [-1, config.max_length, 2])
 
+        clf_token = encoder['_classify_']
+        pool_idx = tf.cast(tf.argmax(tf.cast(tf.equal(X[:, :, 0], clf_token), tf.float32), 1), tf.int32)
+        if explain:
+            flat_x = tf.reshape(X[:,:,:1], [-1, 1])
```
Comment (benleetownsend) [inline finetune/base_models/gpt/featurizer.py]: This logic is gross, but also largely applicable to other models. Factor out.

File: finetune/target_models/classifier.py [hunk 8]
```diff
@@ -74,15 +76,54 @@ def get_eval_fn(cls):
...
+            logits = clf_out["logits"]
+            clf_out["logits"] = logits[: shape[0]]
+            clf_out["explanation"] = tf.nn.softmax(
+                tf.reshape(
+                    logits[shape[0]:], tf.concat((shape[:2], [tf.shape(logits)[-1]]), 0)
+                ),
+            -1)
+        return clf_out
+
+    def explain(self, Xs):
```
Comment (benleetownsend) [inline finetune/target_models/classifier.py]: Document.

File: finetune/base_models/__init__.py (hunk 1 above)
Comment (madisonmay) [reply finetune/base_models/__init__.py]: Just GPT is fine for now

Comment (madisonmay) [reply finetune/base_models/__init__.py]: Oh just the **kwargs -- got it.

File: finetune/base_models/gpt/encoder.py [hunk 9]
```diff
@@ -188,37 +188,47 @@ def decode(self, ids):
...
     spacy_start = None
     for token, prob, start, end in zip(tokens, attn, token_starts, token_ends):
         to_combine.append(prob)
-        if not spacy_start:
+        if spacy_start is None:
             spacy_start = start
         if token.endswith('</w>'):
-            spacy_attn.append(max(to_combine))
+            if attention:
+                spacy_attn.append(np.max(to_combine, 0))
```
Comment (madisonmay) [inline finetune/base_models/gpt/encoder.py]: Maybe rename this variable.

File: finetune/base_models/gpt/encoder.py (hunk 3 above)
Comment (madisonmay) [reply finetune/base_models/gpt/encoder.py]: Only nominally, it's not documented and I'm sure no one uses it.

**Pull request merged**